runtime/src/cost_model.rs

```diff
@@ -153,19 +160,25 @@ impl CostModel {
         data_bytes_cost
     }
 
-    fn get_transaction_cost(&self, transaction: &SanitizedTransaction) -> u64 {
-        let mut cost: u64 = 0;
+    fn get_transaction_cost(&self, transaction: &SanitizedTransaction) -> (u64, u64) {
+        let mut builtin_costs = 0u64;
+        let mut bpf_costs = 0u64;
 
         for (program_id, instruction) in transaction.message().program_instructions_iter() {
-            let instruction_cost = self.find_instruction_cost(program_id);
-            trace!(
-                "instruction {:?} has cost of {}",
-                instruction,
-                instruction_cost
-            );
-            cost = cost.saturating_add(instruction_cost);
+            // to keep the same behavior, look for builtin first
```

Comment:
I think the behavior is actually different with this change. The builtin program entries in the instruction cost table used to get updated with 0 compute units after instruction processing. Now we do the right thing and just use the preset cost.

I would prefer if you fixed that issue in a separate PR that focuses just on getting builtin compute units correct.

Reply:
> The behavior didn't change. Builtin programs' units were added to cost table during table initializing ([here](https://github.com/solana-labs/solana/pull/24311/files#diff-05b5f07ae482624062494888162a725444b51ee716f1f147affef29b785386c7L82)), which is removed. So builtin programs still get their cost from their `BUILT_IN_INSTRUCTION_COSTS` table, they just no longer mingled with BPF programs in cost_table

After they get initialized there's nothing special about them though. I think that they will get updated over time due to the fact that `accumulate_program` doesn't discriminate between bpf and builtin programs here: https://github.com/solana-labs/solana/blob/9052e41b3272cedcf63b063104fea16269a9749e/runtime/src/message_processor.rs#L137

Reply:
I swear there's logic to _not_ to update cost table if `program_timing.accumulated_units == 0`, but I can't find it. Thanks for pointing this out! I'll put a separate PR for splitting builtin cost from BPF to correct the behavior, then rebase this PR on top of that.

Reply:
PR #24468 to separate builtins and BPFs. Will rebase it once approved.

Reply:
rebased